ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -14,215 +14,308 @@
 #define NDEBUG
 #include <debug.h>
 
-/* TYPES *******************************************************************/
+/* DATA **********************************************************************/
+
+KGUARDED_MUTEX PiNotifyTargetDeviceLock;
+KGUARDED_MUTEX PiNotifyHwProfileLock;
+KGUARDED_MUTEX PiNotifyDeviceInterfaceLock;
+
+_Guarded_by_(PiNotifyHwProfileLock)
+LIST_ENTRY PiNotifyHwProfileListHead;
+
+_Guarded_by_(PiNotifyDeviceInterfaceLock)
+LIST_ENTRY PiNotifyDeviceInterfaceListHead;
+
+/* TYPES *********************************************************************/
 
 typedef struct _PNP_NOTIFY_ENTRY
 {
     LIST_ENTRY PnpNotifyList;
-    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
     PVOID Context;
-    UNICODE_STRING Guid;
-    PFILE_OBJECT FileObject;
     PDRIVER_OBJECT DriverObject;
     PDRIVER_NOTIFICATION_CALLBACK_ROUTINE PnpNotificationProc;
+    union
+    {
+        GUID Guid; // for EventCategoryDeviceInterfaceChange
+        struct
+        {
+            PFILE_OBJECT FileObject; // for EventCategoryTargetDeviceChange
+            PDEVICE_OBJECT DeviceObject;
+        };
+    };
+    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
+    UINT8 RefCount;
+    BOOLEAN Deleted;
 } PNP_NOTIFY_ENTRY, *PPNP_NOTIFY_ENTRY;
 
-KGUARDED_MUTEX PnpNotifyListLock;
-LIST_ENTRY PnpNotifyListHead;
-
 /* FUNCTIONS *****************************************************************/
 
+CODE_SEG("INIT")
 VOID
-IopNotifyPlugPlayNotification(
-    IN PDEVICE_OBJECT DeviceObject,
-    IN IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
-    IN LPCGUID Event,
-    IN PVOID EventCategoryData1,
-    IN PVOID EventCategoryData2)
+PiInitializeNotifications(VOID)
 {
-    PPNP_NOTIFY_ENTRY ChangeEntry;
-    PLIST_ENTRY ListEntry;
-    PVOID NotificationStructure;
-    BOOLEAN CallCurrentEntry;
-    UNICODE_STRING GuidString;
-    NTSTATUS Status;
-    PDEVICE_OBJECT EntryDeviceObject = NULL;
+    KeInitializeGuardedMutex(&PiNotifyTargetDeviceLock);
+    KeInitializeGuardedMutex(&PiNotifyHwProfileLock);
+    KeInitializeGuardedMutex(&PiNotifyDeviceInterfaceLock);
+    InitializeListHead(&PiNotifyHwProfileListHead);
+    InitializeListHead(&PiNotifyDeviceInterfaceListHead);
+}
 
-    ASSERT(DeviceObject);
+static
+CODE_SEG("PAGE")
+VOID
+PiDereferencePnpNotifyEntry(
```

Comment:
I know we can't annotate/assert a specific lock required here, since multiple ones are possible. But since they're all guarded mutexes, maybe `_Requires_lock_held_(_Global_critical_region_)` + `ASSERT(KeAreAllApcsDisabled())`?
Same for the Reference function of course.